contribute/style-guides/e2e.md

```diff
@@ -0,0 +1,140 @@
+# End to End test framework
+As our company grows our need for end to end (called e2e from now on) tests have also grown. Today we're using a minimal home grown solution built on top of Cypress for our e2e tests. 
+
+## Basic concepts
+A good introduction to some good e2e practices can be found [here](https://martinfowler.com/bliki/PageObject.html).
+- `Selector`, an unique identifier that is used from the e2e framework to retrieve an element from the Browser 
+- `Page`, an abstraction for an object that contains one or more `Selectors`
+- `Flow`, an abstraction that contains a sequence of actions on one or more `Pages` that can be reused and shared between tests
+
+## Basic example
+Lets start with a simple example with a single selector and for simplicity all examples will be in JSX.
+
+In our example app we have an input that we during our e2e test want to type some text in.
+```jsx harmony
+<div>
+    <input type="text" className="gf-form-input login-form-input"/>
+</div>
+```
+
+We could define a selector using `JQuery` [type selectors](https://api.jquery.com/category/selectors/) with a string like `'.gf-form-input.login-form-input'` but that would be brittle as style changes occur frequently and there is nothing that signalizes to future developers that this input is part of an e2e test.
+
+At Grafana we've decided to use `aria-label` as our preferred way of defining our selectors instead of for instance`data-*` attributes because it also aids in accessibility.
+Lets add an descriptive `aria-label` to our simple example.
+```jsx harmony
+<div>
+    <input type="text" className="gf-form-input login-form-input" aria-label="Username input field"/>
+</div>
+```
+
+Now that we added the `aria-label` we suddenly get more information about this particular field. It's an input field that represents an username but there it's still not really signalizing that it's part of a e2e test.
+
+The next step is to create a `Page` representation in our e2e test framework to glue the test with the real implementation using the `pageFactory` function. For that function we can supply an `url` and `selectors` like the example below.
+```typescript
+export const Login = pageFactory({
+  url: '/login', // used when called from Login.visit()
+  selectors: {
+    username: 'Username input field', // used when called from Login.username().type('Hello World')
+  },
+});
+```
+
+The next step is to add `Login` page to the exported const `Pages` in `packages/grafana-e2e/src/pages/index.ts` so that it appears when we type `e2e.pages` in our IDE.
```

Comment:
```suggestion
The next step is to add the `Login` page to the exported const `Pages` in `packages/grafana-e2e/src/pages/index.ts` so that it appears when we type `e2e.pages` in our IDE.
```